packages/tools/devtools/devtools-core/src/BaseDevtools.ts

```diff
@@ -0,0 +1,543 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import type { IAudience } from "@fluidframework/container-definitions";
+import type { IFluidLoadable } from "@fluidframework/core-interfaces";
+import type { IClient } from "@fluidframework/driver-definitions";
+
+import type { AudienceClientMetadata } from "./AudienceMetadata.js";
+import type { ContainerKey, FluidObjectId, HasContainerKey } from "./CommonInterfaces.js";
+import { ContainerStateChangeKind } from "./Container.js";
+import type { ContainerStateMetadata } from "./ContainerMetadata.js";
+import type { DecomposedContainer } from "./DecomposedContainer.js";
+import type { ContainerDevtoolsFeatureFlags } from "./Features.js";
+import type { IContainerDevtools } from "./IContainerDevtools.js";
+import type { AudienceChangeLogEntry, ConnectionStateChangeLogEntry } from "./Logs.js";
+import {
+	DataVisualizerGraph,
+	type FluidObjectNode,
+	type RootHandleNode,
+	VisualNodeKind,
+	defaultVisualizers,
+} from "./data-visualization/index.js";
+import {
+	AudienceSummary,
+	ContainerDevtoolsFeatures,
+	ContainerStateChange,
+	ContainerStateHistory,
+	DataVisualization,
+	GetAudienceSummary,
+	GetContainerDevtoolsFeatures,
+	GetContainerState,
+	GetDataVisualization,
+	GetRootDataVisualizations,
+	type IDevtoolsMessage,
+	type ISourcedDevtoolsMessage,
+	type InboundHandlers,
+	type MessageLoggingOptions,
+	RootDataVisualizations,
+	handleIncomingWindowMessage,
+	postMessagesToWindow,
+} from "./messaging/index.js";
+
+/**
+ * Abstract base class for devtools implementations.
+ * @remarks Contains shared functionality between {@link ContainerDevtools} and {@link ContainerRuntimeDevtool}.
+ */
+export abstract class BaseDevtools<DevtoolsType extends DecomposedContainer>
```

Comment:
Nit: Our guidelines encourage using the `T` prefix for generic type parameters. I would recommend `TContainer` or something similar.

https://github.com/microsoft/FluidFramework/wiki/Coding-Guidelines#-do-prefix-type-parameters-with-t

Reply:
I thought since we are moving away from `I` for interface, we are also not using `T` for type 😆 . Thanks!